source/NRF52Serial.cpp

```diff
@@ -1,139 +1,349 @@
+    // to achieve the lowest power consumption.
+    nrf_uarte_shorts_disable(p_uarte_, NRF_UARTE_SHORT_ENDRX_STARTRX);
+    nrf_uarte_task_trigger(p_uarte_, NRF_UARTE_TASK_STOPRX);
+    nrf_uarte_event_clear(p_uarte_, NRF_UARTE_EVENT_TXSTOPPED);
+    nrf_uarte_task_trigger(p_uarte_, NRF_UARTE_TASK_STOPTX);
+    while (!nrf_uarte_event_check(p_uarte_, NRF_UARTE_EVENT_TXSTOPPED))
+    {}
 
-    if (p_event->type == NRFX_UARTE_EVT_TX_DONE)
-    {
-        tx_complete = 1;
-        instance->dataTransmitted();
+    nrf_uarte_disable(p_uarte_);
+    nrf_uarte_txrx_pins_disconnect(p_uarte_);
+}
+
+void NRF52Serial::_irqHandler(void *self_)
+{
+    NRF52Serial *self = (NRF52Serial *)self_;
+    NRF_UARTE_Type *p_uarte = self->p_uarte_;
+
+    if (nrf_uarte_event_check(p_uarte, NRF_UARTE_EVENT_RXDRDY)){
+        nrf_uarte_event_clear(p_uarte, NRF_UARTE_EVENT_RXDRDY);
+        self->dataReceivedDMA();
+    }
+    
+    if (nrf_uarte_event_check(p_uarte, NRF_UARTE_EVENT_ENDRX)){
+        nrf_uarte_event_clear(p_uarte, NRF_UARTE_EVENT_ENDRX);
+        self->updateRxBufferAfterENDRX();
+        nrf_uarte_task_trigger(p_uarte, NRF_UARTE_TASK_STARTRX);
+    }else if (nrf_uarte_event_check(p_uarte, NRF_UARTE_EVENT_ERROR)){
+        nrf_uarte_event_clear(p_uarte, NRF_UARTE_EVENT_ERROR);
+        (void) nrf_uarte_errorsrc_get_and_clear(p_uarte);
+        // RX transaction abort
+        nrf_uarte_task_trigger(p_uarte, NRF_UARTE_TASK_STOPRX);
+        self->updateRxBufferAfterENDRX();
+        nrf_uarte_task_trigger(p_uarte, NRF_UARTE_TASK_STARTRX);        
     }
 
-    if (p_event->type == NRFX_UARTE_EVT_RX_DONE)
-    {
-        rx_complete = 1;
-        instance->dataReceived(instance->getc());
+    if (nrf_uarte_event_check(p_uarte, NRF_UARTE_EVENT_RXTO)){
+        nrf_uarte_event_clear(p_uarte, NRF_UARTE_EVENT_RXTO);
+        // nrf_uarte_task_trigger(p_uarte, NRF_UARTE_TASK_FLUSHRX);
+        uint32_t rx_cnt = nrf_uarte_rx_amount_get(p_uarte);
+        while(rx_cnt-- > 0){
+            self->dataReceivedDMA();
+        }
+    }
+
+    if (nrf_uarte_event_check(p_uarte, NRF_UARTE_EVENT_ENDTX)){
+        nrf_uarte_event_clear(p_uarte, NRF_UARTE_EVENT_ENDTX);
+
+        self->is_tx_in_progress_ = false;
+        if(self->txBufferedSize() > 0){
+            self->dataTransmitted();
+        }else{
+            // Transmitter has to be stopped by triggering STOPTX task to achieve
+            // the lowest possible level of the UARTE power consumption.
+            nrf_uarte_task_trigger(p_uarte, NRF_UARTE_TASK_STOPTX);
+        }
+    }
+
+    if (nrf_uarte_event_check(p_uarte, NRF_UARTE_EVENT_TXSTOPPED)){
+        nrf_uarte_event_clear(p_uarte, NRF_UARTE_EVENT_TXSTOPPED);
+        self->is_tx_in_progress_ = false;
     }
 }
 
+
 int NRF52Serial::enableInterrupt(SerialInterruptType t)
 {
-    if (t == RxInterrupt)
-        uart_instance.p_reg->INTENSET |= (UARTE_INTENSET_ENDRX_Msk);
+    if(is_configured_ == false)
```

Comment:
It would be better if all of these if statements are removed and instead trigger a panic if allocation fails in the constructor:

`target_panic(DEVICE_HARDWARE_CONFIGURATION_ERROR);`

Reply:
Done on 9f6a6ce.